### ISSUE TITLE
Non-WebSockets Transport using Linux C# Client Fails

### ISSUE THREAD

**Issue by abbotware** [aspnet/SignalR#2230]
C# <-> C# communication from NetCore on Linux seems to not be working when the transport is not websockets.  I suspect this might be an issue on the linux client side since a windows machine connecting to the same server has no problems.  There are also no problems if the same linux machines connects to IIS Express and web sockets are 'working'.

This sums up the scenarios:

OS | WebSockets (IIS Express) | NonWebSockets (AWS / Elastic Beanstalk) |
--- | --- | ---
|Linux | OK | :-(|
|Windows | OK | OK |

* Versions of Server-Side NuGet Packages:
1.0.0-rc1-final
* Versions of Client-Side NuGet/NPM Packages:
1.0.0-rc1-final
* Are you using the C# client or the JavaScript client:
C# client

* The Server you are using (Kestrel/HttpSysServer/IIS/IIS Express/Azure Web App/etc.): 
IIS in AWS (Elastic Beanstalk) 
* The Operating System on the Server (Windows/Linux/macOS):
Windows / AWS
* The Operating System on the Client (Windows/Linux/macOS):
Linux (ARM processor armhf architecture)

**Comment by BrennanConroy** [comment]
Could you turn on server and client logs to see if there is a reason non websockets isn't "working"?

**Comment by abbotware** [comment]
Linux Client - Initial Connection Log
https://gist.github.com/abbotware/ead806f783454ec272af41d643d442fc

**Comment by anurse** [comment]
The console output you provided has WebSocket connection errors in it, but you said WebSockets were working. How are you "disabling" WebSockets? Are you configuring the client to use SSE or Long Polling or disabling WebSockets on the server?

Regardless of that, the logs seem to show that the connection is functioning properly and that the invocation is being dispatched and completed properly. What is the specific behavior you are seeing that you don't expect? Can you provide code samples that illustrate the issue?

**Comment by abbotware** [comment]
The message flow is more or less as follows:

client -> server (online)

client <- server (input)

client -> server (output)
client -> server (output)
client -> server (output)

... Error 

Here is a log of the full connection 

https://gist.github.com/abbotware/0add10d1e54359fda9cffbc3191937ac

**Comment by anurse** [comment]
> I am not disabling anything

Ok, I was getting confused by your table. I see that the server is on the *bottom* of the table now. So you're saying that AWS is failing to connect with WebSockets and falling back to one of the other transports?

Your log shows:

```
fail: Microsoft.AspNetCore.SignalR.Client.HubConnection[38]
      Received close message with an error: Connection closed with an error.
```

For security reasons, the error message is not sent to the client unless it is a known error. You'll need to **either** provide logs from the server or set the `EnableDetailedErrors` setting on `HubOptions` to `true` to get the detailed error message on the client. You can do that in `ConfigureServices`:

```csharp
services.AddSignalR(options => options.EnableDetailedErrors = true);
```

The log also shows an error with the WebSocket handshake. This makes me suspect something about AWS is causing the WebSocket handshake to be corrupted. It looks like this is the case when using the "Classic Load Balancer" with Elastic Beanstalk. Have you configured the "Classic Load Balancer" or the "Application Load Balancer"? Only the "Application Load Balancer" supports WebSockets according to my quick Google search results: https://aws.amazon.com/about-aws/whats-new/2016/08/aws-elastic-beanstalk-supports-application-load-balancer/

**Comment by abbotware** [comment]
Its an application load balancer - HTTPS externally and HTTP internally.  But as of now there is only 1 web server behind it so there is are requests being round robined / etc.

**Comment by anurse** [comment]
Ok, the "Application Load Balancer" is supposed to work with WebSockets so that shouldn't be a problem.

If you're just using a single server, I can't see anything in the client logs that would be causing the issue. We'd need server logs in order to get more data. I'm not familiar with Elastic Beanstalk, so I don't know if/how you can get logging data out. You can configure ASP.NET Core logging to log to files or even external services/databases through the help of other third-party components, but I don't know of any specific components to suggest.

**Comment by abbotware** [comment]
https://gist.github.com/abbotware/e679750007ea5ea9b609076c1c9e0611

'InvalidDataException: Missing required property 'type'.'

**Comment by anurse** [comment]
Something between the client and the server seems to be corrupting the messages. Can you collect network traces from the client? See the [Diagnostics Guide](https://github.com/aspnet/SignalR/wiki/Diagnostics-Guide) for some guidance on how to collect this data.

And getting server logs, as I posted above, would help a lot as well.

**Comment by abbotware** [comment]
for server logs - i can think of some creative ways to get log data out - but is there anything i need to do to enable server signalr to log more data on thes server side?

**Comment by abbotware** [comment]
Also - I think can bypass the EB (HTTPS) and go straight to

**Comment by abbotware** [comment]
er whoops - misclick there

**Comment by abbotware** [comment]
I added Serilog to my web app since asp net core doesnt seem to have a built in file logger... attempted to serve the log file from an MVC Action, but looks like the iis file system is locked down so i doesnt create files... i wound up trying to write logs to cloudwatch - but they don't look any where as detailed as the client logs

**Comment by abbotware** [comment]
`public static int Main(string[] args)
        {
            // https://mitchelsellers.com/blogs/2017/10/09/real-world-aspnet-core-logging-configuration

            ////var currentEnv = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");

            var configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json")
                ////.AddJsonFile($"appsettings.{currentEnv}.json", optional: true)
                .AddEnvironmentVariables()
                .Build();

            // name of the log group
            var logGroupName = "XXXXX.XX";

            // customer renderer (optional, defaults to a simple rendered message of Serilog's LogEvent
            //var renderer = new MyCustomRenderer();

            // options for the sink, specifying the log group name
            CloudWatchSinkOptions options = new CloudWatchSinkOptions { LogGroupName = logGroupName};

            // setup AWS CloudWatch client
            AWSCredentials credentials = new BasicAWSCredentials("XXXXXXXXXXXXXX", "XXXXXXXXXXXXXXXXXXXXXXXXXX");
            IAmazonCloudWatchLogs client = new AmazonCloudWatchLogsClient(credentials, RegionEndpoint.USEast1);

            Log.Logger = new LoggerConfiguration()
                .ReadFrom.Configuration(configuration)
                .MinimumLevel.Verbose()
                .Enrich.FromLogContext()
                .WriteTo.Console()
                .WriteTo.AmazonCloudWatch(options, client)
                .WriteTo.RollingFile("serilog.txt", outputTemplate: "{Timestamp} [{Level}] ({CategoryName:l}) {Message}{NewLine}{Exception}", shared: true)
                .CreateLogger();

            try
            {
                BuildWebHost(args).Run();

                return 0;
            }
            catch (Exception ex)
            {
                Log.Fatal(ex, "Host terminated unexpectedly");
                return 1;
            }
            finally
            {
                Log.CloseAndFlush();
            }
        }

        public static IWebHost BuildWebHost(string[] args)
        {
            return WebHost.CreateDefaultBuilder(args)
                .UseKestrel()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseIISIntegration()
                .UseStartup<Startup>()
                .UseApplicationInsights()
                .UseSetting("detailedErrors", "true")
                .CaptureStartupErrors(true)
                .ConfigureLogging(logging => {
                    logging.AddFilter("Microsoft.AspNetCore.SignalR", LogLevel.Trace);
                    logging.AddFilter("Microsoft.AspNetCore.Http.Connections", LogLevel.Trace);
                })
                .UseSerilog()
                .Build();
        }`

**Comment by abbotware** [comment]
am I doing something wrong?  the log data i am getting seems extremely limited...   nothing hub / connection releated

**Comment by davidfowl** [comment]
Do you have an appsettings.json? That will override what you have configured in code.

**Comment by abbotware** [comment]
i do but here is the strange part - in my serilog file i see debug and verbose... but cloudwatch does not have any debug or verbose...

**Comment by abbotware** [comment]
appsettings.json values: 

  "Logging": {
    "IncludeScopes": true,
    "LogLevel": {
      "Default": "Debug",
      "System": "Information",
      "Microsoft": "Information",
      "Microsoft.AspNetCore.SignalR": "Trace",
      "Microsoft.AspNetCore.Http.Connections": "Trace"
    }
  }

**Comment by abbotware** [comment]
https://gist.github.com/abbotware/1bd88d6ba631df3cdc1aeb18d140478e

doesnt look like too much in the log:   "Error when processing requests." 

the log format looks like its missing columns of data (ie level, category, etc)

**Comment by davidfowl** [comment]
Can you get the console logger output? That error should have the actual exception associated with it but for some reason it's not showing up for you.

**Comment by abbotware** [comment]
I can easily reproduce this and im at a point where I can get all logging data out of the system, so if you need me to test / change anything I can

**Comment by abbotware** [comment]
If I bypass the HTTPS to HTTP Load Balancer and connect directly to the web server - same problem - Also, I am kinda surprised that WebSockets are not being used when i connect directly to the HTTP server...  does app.UseWebSockets need to be specified in addition to app.UseSignalR ?

**Comment by davidfowl** [comment]
If you can reproduce this on demand I'd like to take a look at debugging this with you.

**Comment by davidfowl** [comment]
@anurse for this error:

```
[5/17/2018 1:21:15 AM +00:00] [Error] Error when processing requests. System.IO.InvalidDataException: Missing required property 'type'.
at Microsoft.AspNetCore.SignalR.Protocol.JsonHubProtocol.ParseMessage(Utf8BufferTextReader textReader, IInvocationBinder binder)
at Microsoft.AspNetCore.SignalR.Protocol.JsonHubProtocol.TryParseMessage(ReadOnlySequence1& input, IInvocationBinder binder, HubMessage& message) at Microsoft.AspNetCore.SignalR.HubConnectionHandler1.d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.d__11.MoveNext()
```

We have a way to look at the payload on the server right? 

@abbotware can you turn trace logging on the client? It looks like something is sending messages out of order or sending a bad payload.

**Comment by abbotware** [comment]
the full client log is above:  https://gist.github.com/abbotware/0add10d1e54359fda9cffbc3191937ac

I am going to test this out an a regular IIS server (non-aws) to see if I can repro there... if yes, then i can get a lot more server diagnostics

**Comment by davidfowl** [comment]
There are 2 things to figure out here:

1. What is messing with the websocket request causing this error:

```C#
System.Net.WebSockets.WebSocketException (0x80004005): The 'Sec-WebSocket-Accept' header value '3VpOLaVXO6aI9Z7GUGxzkt4v2ws=' is invalid.
   at System.Net.WebSockets.WebSocketHandle.ValidateAndTrackHeader(String targetHeaderName, String targetHeaderValue, String foundHeaderName, String foundHeaderValue, Boolean& foundHeader)
   at System.Net.WebSockets.WebSocketHandle.<ParseAndValidateConnectResponseAsync>d__28.MoveNext()
```

I see you are using IIS in AWS (Elastic Beanstalk). Do you have the WebSocket module enabled on IIS there?

I saw this http://blog.flux7.com/web-apps-websockets-with-aws-elastic-load-balancing but I don't know how relevant it is to your deployment.

2. Why is it barfing after a while. We need to get the message content. I'm pretty sure the server logs the message content but that might only be on a successful parse 😢 

/cc @Tratcher

**Comment by Tratcher** [comment]
Do you have a capture of the request headers? It's easy enough to reproduce and verify the Sec-WebSocket-Accept value.
https://github.com/aspnet/WebSockets/blob/8513fc832943454c15e0138bb7b278f0b2c84c2c/src/Microsoft.AspNetCore.WebSockets/Internal/HandshakeHelpers.cs#L99-L118

**Comment by abbotware** [comment]
hey  so im still trying to get more info but regarding AWS EB - I did nothing to specifically enable websockets on the load balancer or EB.. I am only using HTTPS -> HTTP 'load balancer' to put an SSL connecting in front of a single server 'bean stalk'

**new update**: if I by pass the Load Balancer and go directly to the single EB node - I get the same problem - I would say this rules out the 'load balancer'

Keep in mind, in ALL cases, a windows client connecting to the same server has no problems! (hence why i think this is an issue on the linux client) (forgot to mention, but its an **ARM** processor if that is of any significance)

**Comment by abbotware** [comment]
ok - I reproduced the issue outside of AWS on a windows server running IIS and captured the communication via Wireshark.  I tried to filter out as much as i could so its mostly the 
server: 10.10.10.30 and client: 192.168.2.222

**Comment by davidfowl** [comment]
Very nice! This IIS Server has websockets enabled right? It must if windows just works. Are you getting the same websocket error?

**Comment by abbotware** [comment]
well no, Its IIS and its failing back to non websockets just like AWS - so all the client and server logs i have posted probably match this capture sequence as well..     The only websockets i have seen working was running asp net core locally via IIS express.

**Comment by abbotware** [comment]
curious - It looks like non-websockets uses 2 separate HTTP streams? (input / output) ?

**Comment by davidfowl** [comment]
> well no, Its IIS and its failing back to non websockets just like AWS - so all the client and server logs i have posted probably match this capture sequence as well.. The only websockets i have seen working was running asp net core locally via IIS express.

OK so that explains this then:

> What is messing with the websocket request causing this error:

It's because you haven't turned on WebSockets on IIS. See https://docs.microsoft.com/en-us/iis/get-started/whats-new-in-iis-8/iis-80-websocket-protocol-support for how to turn it on.

> curious - It looks like non-websockets uses 2 separate HTTP streams? (input / output) ? or am i just seeing 2 seperate streams on one connection in the capture?

Yes there are 2 different requests for input and output for both Server-Sent Events and Long-Polling.

**Comment by abbotware** [comment]
well yes i figured I could turn web sockets on, but - there is a some bug here no?

**Comment by davidfowl** [comment]
> well yes i figured I could turn web sockets on, but - there is some bug here no?

I'm not denying there's a bug. Remember here https://github.com/aspnet/SignalR/issues/2230#issuecomment-390108893, I was trying to clarify 2 things but it looks as though we can ignore why websockets isn't working and focus on the other problem.

**Comment by davidfowl** [comment]
One thing I forgot to ask, @abbotware are you using the .NET Client on .NET Core 2.1 or something else?

**Comment by abbotware** [comment]
Net Core 2.0 for both server and client

**Comment by abbotware** [comment]
If it were to help - I might be able to extract all the SignalR code I am using and make a small client / server sample project - however - i'd prefer not to post it online

**Comment by BrennanConroy** [comment]
That would be helpful, you can email me the sample at 'brecon at microsoft dot com'

**Comment by davidfowl** [comment]
Can you try updating to .NET Core 2.1? At least on the client. The HttpClient implementation is vastly improved and I wonder if that's the culprit here. 

You can try out the early access drop here https://github.com/aspnet/Home/wiki/2.1.0-Early-Access-Downloads

**Comment by davidfowl** [comment]
@abbotware can you also capture the server logs? Or if you already posted those can you link us to them?

**Comment by abbotware** [comment]
here is a summary of all the logs / data

client:   https://gist.github.com/abbotware/0add10d1e54359fda9cffbc3191937ac

server:  https://gist.github.com/abbotware/1bd88d6ba631df3cdc1aeb18d140478e

i was able to grab the following exception which is not in the server log posted but from another log:

[5/17/2018 1:21:15 AM +00:00] [Error] Error when processing requests. System.IO.InvalidDataException: Missing required property 'type'.
at Microsoft.AspNetCore.SignalR.Protocol.JsonHubProtocol.ParseMessage(Utf8BufferTextReader textReader, IInvocationBinder binder)
at Microsoft.AspNetCore.SignalR.Protocol.JsonHubProtocol.TryParseMessage(ReadOnlySequence1& input, IInvocationBinder binder, HubMessage& message) at Microsoft.AspNetCore.SignalR.HubConnectionHandler1.d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.d__11.MoveNext()

in all cases the message exchange is as follows:

**Comment by abbotware** [comment]
hmm i never installed a 'pre-release' 'final' build,,,  :-)  so my build / nuget restore step seems confused:

 error NU1102: Unable to find package runtime.win-x64.Microsoft.NETCore.App with version (>= 2.1.0)
C:\go\pipelines\Trunk\Zyronet\Zyronet.Host.Task\Zyronet.Host.Task.csproj : error NU1102:   - Found 11 version(s) in nuget.org [ Nearest version: 2.1.0-rc1 ]

i mean at th

**Comment by BrennanConroy** [comment]
From the Wireshark logs I'm leaning towards it being a client side issue. You're using Linux Arm32 right? Are you able to consistently repro with the same amount of sends? The server and wireshark logs show it failed after sending "Abbotware.Interop.Microsoft.dll"

**Comment by abbotware** [comment]
yes i can repro all the time - all the logs are from separate runs - so can't be just a coincidence  - I havent really looked at the number of sends being made though

**Comment by BrennanConroy** [comment]
Official Arm32 support on Linux was added in 2.1.0. So it's very likely this is a 2.0 Arm32 specific bug that's been fixed in 2.1.

How did your production testing go?

**Comment by abbotware** [comment]
Upgraded everything now to net core 2.1 (client and server) and haven't seen this issue locally or remote (AWS).  So guess its probably safe to close - Thanks everyone for the help :-)
